File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
import java.util.PriorityQueue;

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {

        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);

        // Put the first node of every list into the heap
        for (ListNode node : lists) {
            if (node != null) {
                pq.add(node);
            }
        }

        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;

        while (!pq.isEmpty()) {

            ListNode temp = pq.poll(); // Smallest node

            curr.next = temp;
            curr = curr.next;

            if (temp.next != null) {
                pq.add(temp.next);
            }
        }

        return dummy.next;
    }
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
import heapq

class Solution:
    def mergeKLists(self, lists):

        heap = []

        # Add the first node of every list
        for i, node in enumerate(lists):
            if node:
                heapq.heappush(heap, (node.val, i, node))

        dummy = ListNode(-1)
        curr = dummy

        while heap:

            val, i, node = heapq.heappop(heap)

            curr.next = node
            curr = curr.next

            if node.next:
                heapq.heappush(heap, (node.next.val, i, node.next))

        return dummy.next